apps/prairielearn/src/ee/pages/administratorInstitutionSaml/administratorInstitutionSaml.ts

```diff
@@ -75,38 +102,49 @@ router.post(
           authn_user_id: res.locals.authn_user.user_id,
         });
       });
+      res.redirect(req.originalUrl);
     } else if (req.body.__action === 'delete') {
       await queryAsync(sql.delete_institution_saml_provider, {
         institution_id: req.params.institution_id,
         // For audit logs
         authn_user_id: res.locals.authn_user.user_id,
       });
-    } else {
-      throw new error.HttpStatusError(400, `unknown __action: ${req.body.__action}`);
-    }
+      res.redirect(req.originalUrl);
+    } else if (req.body.__action === 'decodeAssertion') {
+      const samlConfig = await getSamlOptions({
+        institution_id: req.params.institution_id,
+        host: req.headers.host,
+        strictMode: req.body.strict_mode === '1',
+      });
+      const saml = new SAML({
+        ...samlConfig,
+        // Disable clock skew checking; we might be testing with a very old SAML response.
+        acceptedClockSkewMs: -1,
+      });
 
-    res.redirect(req.originalUrl);
-  }),
-);
+      let xml: string;
+      try {
+        xml = formatXml(Buffer.from(req.body.encoded_assertion, 'base64').toString('utf8'));
+      } catch (err) {
+        res.send(DecodedAssertion({ xml: err.message, profile: '' }));
+        return;
+      }
 
-router.get(
-  '/',
-  asyncHandler(async (req, res) => {
-    const institution = await getInstitution(req.params.institution_id);
-    const samlProvider = await getInstitutionSamlProvider(req.params.institution_id);
-    const institutionAuthenticationProviders = await getInstitutionAuthenticationProviders(
-      req.params.institution_id,
-    );
+      const profile = await saml
+        .validatePostResponseAsync({
+          SAMLResponse: req.body.encoded_assertion,
+        })
+        .catch((err) => {
+          console.error(err);
+          return {
+            error: err.message,
+          };
+        });
```

Comment:
Is there a reason that you are using try/catch for `xml` above, but `.catch()` for `profile` here?

Reply:
`.catch()` means I don't have to do the weird little dance of declaring a variable outside the `try`/`catch`. I prefer to avoid that when possible.